src/linux.rs

```diff
@@ -0,0 +1,275 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{
+    ffi::CStr,
+    io::{Error, ErrorKind},
+    mem,
+    net::IpAddr,
+    os::fd::{AsFd, AsRawFd, BorrowedFd, FromRawFd, OwnedFd},
+    ptr,
+};
+
+use libc::{
+    c_int, c_uchar, c_uint, c_ushort, nlmsghdr, read, socket, write, AF_INET, AF_INET6, AF_NETLINK,
+    AF_UNSPEC, ARPHRD_NONE, IFLA_IFNAME, IFLA_MTU, NETLINK_ROUTE, NLM_F_ACK, NLM_F_REQUEST,
+    RTA_DST, RTA_OIF, RTM_GETLINK, RTM_GETROUTE, RTM_NEWLINK, RTM_NEWROUTE, RTN_UNICAST,
+    RT_SCOPE_UNIVERSE, RT_TABLE_MAIN, SOCK_RAW,
+};
+
+use crate::{default_err, next_item_aligned_by_four};
+
+const NETLINK_BUFFER_SIZE: usize = 8192; // See netlink(7) man page.
+
+#[allow(non_camel_case_types, clippy::struct_field_names)]
+#[repr(C)]
+struct ifinfomsg {
+    ifi_family: c_uchar, // AF_UNSPEC
+    ifi_type: c_ushort,  // Device type
+    ifi_index: c_int,    // Interface index
+    ifi_flags: c_uint,   // Device flags
+    ifi_change: c_uint,  // change mask
+}
+
+#[allow(non_camel_case_types, clippy::struct_field_names)]
+#[repr(C)]
+struct rtmsg {
+    rtm_family: c_uchar,   // Address family of route
+    rtm_dst_len: c_uchar,  // Length of destination
+    rtm_src_len: c_uchar,  // Length of source
+    rtm_tos: c_uchar,      // TOS filter
+    rtm_table: c_uchar,    // Routing table ID; see RTA_TABLE below
+    rtm_protocol: c_uchar, // Routing protocol; see below
+    rtm_scope: c_uchar,    // See below
+    rtm_type: c_uchar,     // See below
+    rtm_flags: c_uint,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+struct rtattr {
+    rta_len: c_ushort,  // Length of option
+    rta_type: c_ushort, // Type of option
+} // Data follows
+
+const fn addr_len(remote: &IpAddr) -> u8 {
+    match remote {
+        IpAddr::V4(_) => 32,
+        IpAddr::V6(_) => 128,
+    }
+}
+
+fn addr_bytes(remote: &IpAddr) -> Vec<u8> {
+    match remote {
+        IpAddr::V4(ip) => ip.octets().to_vec(),
+        IpAddr::V6(ip) => ip.octets().to_vec(),
+    }
+}
+
+fn prepare_nlmsg(
+    nlmsg_type: c_ushort,
+    nlmsg_len: usize,
+    nlmsg_seq: u32,
+) -> Result<nlmsghdr, Error> {
+    let mut nlm = unsafe { mem::zeroed::<nlmsghdr>() };
+    nlm.nlmsg_len = nlmsg_len.try_into().map_err(|_| default_err())?;
+    nlm.nlmsg_type = nlmsg_type;
+    nlm.nlmsg_flags = (NLM_F_REQUEST | NLM_F_ACK)
+        .try_into()
+        .map_err(|_| default_err())?;
+    nlm.nlmsg_seq = nlmsg_seq;
+    Ok(nlm)
```

Comment:
```suggestion
    let nlm = nlmsghdr {
        nlmsg_len: nlmsg_len.try_into().map_err(|_| default_err())?,
        nlmsg_type,
        nlmsg_flags: (NLM_F_REQUEST | NLM_F_ACK)
            .try_into()
            .map_err(|_| default_err())?,
        nlmsg_seq,
        nlmsg_pid: 0,
    };
    Ok(nlm)
```

Would Rust style struct initialization be unsafe here?

Reply:
Yes, the sanitizer complains on `memcpy` if any padding wasn't initialized.